src/auslib/util/cache.py

```diff
@@ -98,3 +121,99 @@ def invalidate(self, name, key):
 
         self.caches[name].invalidate(key)
+
+
+class RedisCache:
+    """A thin wrapper around the redis client to expose a similar interface
+    as ExpiringLRUCache. Unlike ExpiringLRUCache, redis does not support
+    non-trivial objects, so objects are pickled before storage and unpickled
+    upon retrieval.
+
+    This cache can be used on its own, but ideally it is only used through a
+    TwoLayerCache (see below).
+    """
+
+    def __init__(self, redis, name, timeout):
+        self._name = name
+        self._redis = redis
+        # redis and repoze calculate expiry slightly differently; a timeout of
+        # 5 seconds with repoze ends up being 6 seconds in redis. this really
+        # doesn't matter...but it's better to be consistent than not, and redis'
+        # behaviour is slightly confusing, so we make this small improvement
+        # since we're wrapping it anyways.
+        self._timeout = timeout - 1
+        self.lookups = 0
+        self.hits = 0
+        self.misses = 0
+
+    def fullkey(self, key):
+        return f"{self._name}-{key}"
+
+    def get(self, key, default=None):
+        self.lookups += 1
+        if value := self._redis.get(self.fullkey(key)):
+            self.hits += 1
+            return pickle.loads(value)
+
+        self.misses += 1
+        return default
+
+    def put(self, key, value):
+        self._redis.setex(self.fullkey(key), self._timeout, pickle.dumps(value))
+
+    def clear(self):
+        self._redis.delete(*self._redis.keys(self._name))
+
+    def invalidate(self, key):
+        self._redis.delete(self.fullkey(key))
+
+    def remaining_timeout(self, key):
+        absolute_timeout = self._redis.expiretime(self.fullkey(key))
+        return absolute_timeout - time.time()
+
+
+class TwoLayerCache:
+    """A cache that wraps both a RedisCache and ExpiringLRUCache. The
+    former is treated as authoritative, while the latter is used to minimize
+    unnecessary fetches from Redis. This design allows caches to be shared
+    across many pods while minimizing the perf impact of having an off-machine
+    cache."""
+
+    def __init__(self, redis, name, maxsize, timeout):
+        self._redis_cache = RedisCache(redis, name, timeout)
+        self._lru_cache = ExpiringLRUCache(maxsize, timeout)
+        self.lookups = 0
+        self.hits = 0
+        self.misses = 0
+
+    def get(self, key, default=None):
+        self.lookups += 1
+        value = self._lru_cache.get(key, default)
+        if value == default:
+            value = self._redis_cache.get(key, default)
+            # ensure the LRU cache timeout matches the one in redis
+            # this is important to ensure that when new pods spin up that
+            # they will not cache keys for longer than redis
+            # this ensures that we don't have to wait for multiple caches
+            # to expire to refresh data.
+            self._lru_cache.put(key, value, self._redis_cache.remaining_timeout(key))
```

Comment:
I guess there's a small window here where the key might be expired in redis but not the lru cache (since `time.time()` is called first in `remaining_timeout` and then in `ExpiringLRUCache.put`). We're also assuming clocks are synchronized.
(I think that's OK, just figured I'd call it out.)

Reply:
That's a good point, yeah. In an ideal world I guess we'd want to be able to pass in an absolute timestamp for the timeout to avoid this, but `ExpiringLRUCache` doesn't support that.

Presumably these will be close enough in the vast majority of cases; I'll add a comment to this effect though.